src/Microsoft.TemplateEngine.Edge/Template/ParameterSetBuilder.cs

```diff
@@ -0,0 +1,361 @@
+                }
+            }
+            // This for newly optional that does not have value set
+            else if (!isRequired && parameter.DefaultValue != null)
+            {
+                object? resolvedValue = generator.ConvertParameterValueToType(environment, parameter, parameter.DefaultValue, out bool valueResolutionError);
+                if (!valueResolutionError)
+                {
+                    if (resolvedValue is null)
+                    {
+                        throw new InvalidOperationException($"{nameof(resolvedValue)} cannot be null when {nameof(valueResolutionError)} is 'false'.");
+                    }
+                    templateParams.SetParameterValue(parameter, resolvedValue, DataSource.Default);
+                }
+                else
+                {
+                    paramsWithInvalidValues.Add(parameter.Name);
+                }
+            }
+        }
+
+        private static void RunDatasetEvaluation(List<EvalData> evaluatedParameters, IGenerator generator, ILogger logger)
+        {
+            Dictionary<string, EvalData> variables =
+                evaluatedParameters
+                    .Where(p => p.Value != null)
+                    .ToDictionary(p => p.ParameterDefinition.Name, p => p);
+
+            IDictionary<string, object> variableCollection =
+                variables.ToDictionary(p => p.Key, p => p.Value.Value!);
+
+            EvaluateEnablementConditions(generator, evaluatedParameters, variableCollection, variables, logger);
+            EvaluateRequirementCondition(generator, evaluatedParameters, variableCollection, logger);
+        }
+
+        private static void EvaluateEnablementConditions(
+            IGenerator generator,
+            IReadOnlyList<EvalData> parameters,
+            IDictionary<string, object> variableCollection,
+            Dictionary<string, EvalData> variables,
+            ILogger logger)
+        {
+            Dictionary<EvalData, HashSet<EvalData>> parametersDependencies = new();
+
+            // First parameters traversal.
+            //   - evaluate all IsEnabledCondition - and get the dependecies between the parameters during doing so
+            foreach (EvalData parameter in parameters)
+            {
+                if (!string.IsNullOrEmpty(parameter.ParameterDefinition.Precedence.IsEnabledCondition))
+                {
+                    HashSet<string> referencedVariablesKeys = new HashSet<string>();
+                    // Do not remove from the variable collection though - we want to capture all dependencies between parameters in the first traversal.
+                    // Those will be bulk removed before second traversal (traversing only the required dependencies).
+                    parameter.IsEnabledConditionResult =
+                        generator.EvaluateFromString(logger, parameter.ParameterDefinition.Precedence.IsEnabledCondition!, variableCollection, referencedVariablesKeys);
```

Comment:
do we want a better method here? something like `TryEvaluateFromString(... out result, out unknownVars)`.